Stop watching file system if watcher thread quits unexpectedly

If the watcher thread exits due to an exception, we log the following, but do not disable watching:

```text
2021-04-07T08:53:51.765+0200 [ERROR] [system.err] Exception in thread "File watcher consumer" java.lang.RuntimeException: Error!!!
2021-04-07T08:53:51.765+0200 [ERROR] [system.err] 	at org.gradle.internal.watch.registry.impl.DefaultFileWatcherRegistry.lambda$createAndStartEventConsumerThread$0(DefaultFileWatcherRegistry.java:83)
2021-04-07T08:53:51.765+0200 [ERROR] [system.err] 	at java.base/java.lang.Thread.run(Thread.java:834)
```

We should probably do that.

---
cc: @gradle/execution

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/FileWatcherRegistry.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/FileWatcherRegistry.java
@@ -41,7 +41,8 @@ public interface FileWatcherRegistry extends Closeable {
         CREATED,
         MODIFIED,
         REMOVED,
-        INVALIDATED
+        INVALIDATED,
+        OVERFLOW
     }
 
     /**

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
@@ -100,11 +100,11 @@ public class DefaultFileWatcherRegistry implements FileWatcherRegistry {
                                 if (absolutePath == null) {
                                     LOGGER.info("Overflow detected (type: {}), invalidating all watched hierarchies", type);
                                     for (Path watchedHierarchy : fileWatcherUpdater.getWatchedHierarchies()) {
-                                        handler.handleChange(INVALIDATED, watchedHierarchy);
+                                        handler.handleChange(OVERFLOW, watchedHierarchy);
                                     }
                                 } else {
                                     LOGGER.info("Overflow detected (type: {}) for watched path '{}', invalidating", type, absolutePath);
-                                    handler.handleChange(INVALIDATED, Paths.get(absolutePath));
+                                    handler.handleChange(OVERFLOW, Paths.get(absolutePath));
                                 }
                             }
 

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
@@ -44,6 +44,7 @@ import java.util.logging.Level;
 import static org.gradle.internal.watch.registry.FileWatcherRegistry.Type.CREATED;
 import static org.gradle.internal.watch.registry.FileWatcherRegistry.Type.INVALIDATED;
 import static org.gradle.internal.watch.registry.FileWatcherRegistry.Type.MODIFIED;
+import static org.gradle.internal.watch.registry.FileWatcherRegistry.Type.OVERFLOW;
 import static org.gradle.internal.watch.registry.FileWatcherRegistry.Type.REMOVED;
 
 public class DefaultFileWatcherRegistry implements FileWatcherRegistry {
@@ -97,12 +98,12 @@ public class DefaultFileWatcherRegistry implements FileWatcherRegistry {
                             @Override
                             public void handleOverflow(FileWatchEvent.OverflowType type, @Nullable String absolutePath) {
                                 if (absolutePath == null) {
-                                    LOGGER.info("Overflow detected, invalidating all watched hierarchies");
+                                    LOGGER.info("Overflow detected (type: {}), invalidating all watched hierarchies", type);
                                     for (Path watchedHierarchy : fileWatcherUpdater.getWatchedHierarchies()) {
                                         handler.handleChange(INVALIDATED, watchedHierarchy);
                                     }
                                 } else {
-                                    LOGGER.info("Overflow detected for watched hierarchy '{}', invalidating", absolutePath);
+                                    LOGGER.info("Overflow detected (type: {}) for watched path '{}', invalidating", type, absolutePath);
                                     handler.handleChange(INVALIDATED, Paths.get(absolutePath));
                                 }
                             }

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
@@ -131,7 +131,7 @@ public class DefaultFileWatcherRegistry implements FileWatcherRegistry {
         thread.setDaemon(true);
         thread.setName("File watcher consumer");
         thread.setUncaughtExceptionHandler((failedThread, exception) -> {
-            LOGGER.error("File watching stopped due to exception", exception);
+            LOGGER.error("File system event consumer thread stopped due to exception", exception);
             handler.stopWatchingAfterError();
         });
         thread.start();

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
@@ -129,6 +129,10 @@ public class DefaultFileWatcherRegistry implements FileWatcherRegistry {
         });
         thread.setDaemon(true);
         thread.setName("File watcher consumer");
+        thread.setUncaughtExceptionHandler((failedThread, exception) -> {
+            LOGGER.error("File watching stopped due to exception", exception);
+            handler.stopWatchingAfterError();
+        });
         thread.start();
         return thread;
     }

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherRegistry.java
@@ -75,6 +75,7 @@ public class DefaultFileWatcherRegistry implements FileWatcherRegistry {
 
     private Thread createAndStartEventConsumerThread(ChangeHandler handler) {
         Thread thread = new Thread(() -> {
+            LOGGER.debug("Started listening to file system change events");
             try {
                 while (consumeEvents) {
                     FileWatchEvent nextEvent = fileEvents.take();
@@ -124,6 +125,7 @@ public class DefaultFileWatcherRegistry implements FileWatcherRegistry {
                 Thread.currentThread().interrupt();
                 // stop thread
             }
+            LOGGER.debug("Finished listening to file system change events");
         });
         thread.setDaemon(true);
         thread.setName("File watcher consumer");

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java
@@ -350,6 +350,7 @@ public class WatchingVirtualFileSystem extends AbstractVirtualFileSystem impleme
     }
 
     private SnapshotHierarchy stopWatchingAndInvalidateHierarchy(SnapshotHierarchy currentRoot) {
+        LOGGER.error("Stopping file watching and invalidating VFS after an error happened");
         if (watchRegistry != null) {
             try {
                 FileWatcherRegistry toBeClosed = watchRegistry;

--- a/subprojects/soak/src/integTest/groovy/org/gradle/vfs/FileSystemWatchingSoakTest.groovy
+++ b/subprojects/soak/src/integTest/groovy/org/gradle/vfs/FileSystemWatchingSoakTest.groovy
@@ -154,7 +154,7 @@ class FileSystemWatchingSoakTest extends DaemonIntegrationSpec implements FileSy
     }
 
     private static boolean detectOverflow(DaemonFixture daemon, long fromLine) {
-        boolean overflowDetected = daemon.logContains(fromLine, FileWatcherRegistry.Type.INVALIDATED.toString())
+        boolean overflowDetected = daemon.logContains(fromLine, FileWatcherRegistry.Type.OVERFLOW.toString())
         if (overflowDetected) {
             println "Detected overflow in watcher, no files will be retained for the next build"
         }

